File: src/model/ipc/commands.ts
import { PAD_COUNT, type PadId, type PadMode } from "@/model/domain/pad";
import type { Theme } from "@/model/domain/theme";
import { ipc } from "./client";

/** Typed wrappers over the Rust `#[tauri::command]`s. */
export const triggerPad = (pad: PadId): Promise<void> =>
  ipc.invoke("trigger_pad", { pad });

export const releasePad = (pad: PadId): Promise<void> =>
  ipc.invoke("release_pad", { pad });

export const setPadMode = (pad: PadId, mode: PadMode): Promise<void> =>
  ipc.invoke("set_pad_mode", { pad, mode });

/** Apply one play mode to every pad (the global mode selector). */
export const setAllPadModes = (mode: PadMode): void => {
  for (let pad = 0; pad < PAD_COUNT; pad++) void setPadMode(pad, mode);
};

/** Load (or replace) a single pad's sound from a file on disk. */
export const loadPadSound = (pad: PadId, path: string): Promise<void> =>
  ipc.invoke("load_pad_sound", { pad, path });

/** A pad filled by a `.beat` pack, with the mode the pack assigned it. */
export interface LoadedPad {
  pad: PadId;
  mode: PadMode;
}

/** Load a `.beat` pack, replacing the board. Resolves with the filled pads. */
export const loadBeatPack = (path: string): Promise<LoadedPad[]> =>
  ipc.invoke<LoadedPad[]>("load_beat_pack", { path });

/** Remove a single pad's sound. */
export const clearPad = (pad: PadId): Promise<void> =>
  ipc.invoke("clear_pad", { pad });

/** Stop every running loop at once (panic). */
export const stopAll = (): Promise<void> => ipc.invoke("stop_all");

/** Manifest fields supplied when saving a board to a `.beat`. */
export interface SaveMeta {
  name: string;
  author: string;
  bpm: number;
}

/** Write the current board to a `.beat` archive. `modes` is indexed by pad. */
export const saveBeat = (
  path: string,
  meta: SaveMeta,
  modes: PadMode[],
): Promise<void> => ipc.invoke("save_beat", { path, meta, modes });

/** The OS account login name — the default (editable) pack author. */
export const systemUsername = (): Promise<string> =>
  ipc.invoke<string>("system_username");

/** Show or hide the native menu bar (the UI owns + persists the state). */
export const setMenuBarVisible = (visible: boolean): Promise<void> =>
  ipc.invoke("set_menu_visible", { visible });

/** Enable/disable the board-dependent menu items (File → Close, Edit → Clear). */
export const setBoardEnabled = (enabled: boolean): Promise<void> =>
  ipc.invoke("set_board_enabled", { enabled });

/** Payload of the `loop-changed` event the engine emits when a pad loops/stops. */
export interface LoopChanged {
  pad: PadId;
  looping: boolean;
}

export const onLoopChanged = (handler: (e: LoopChanged) => void) =>
  ipc.listen<LoopChanged>("loop-changed", handler);

/** Payload of `sound-changed`: a pad gained or lost a loaded sample. */
export interface SoundChanged {
  pad: PadId;
  loaded: boolean;
}

export const onSoundChanged = (handler: (e: SoundChanged) => void) =>
  ipc.listen<SoundChanged>("sound-changed", handler);

/** Payload of `sound-load-failed`: a sample failed to decode on the audio thread. */
export interface SoundLoadFailed {
  pad: PadId;
  error: string;
}

export const onSoundLoadFailed = (handler: (e: SoundLoadFailed) => void) =>
  ipc.listen<SoundLoadFailed>("sound-load-failed", handler);

/** Result of an update check (mirrors the Rust `UpdateStatus`). */
export type UpdateStatus =
  | { kind: "available"; latest: string }
  | { kind: "upToDate"; current: string }
  | { kind: "failed" };

/** Ask GitHub whether a newer release exists (runs off the UI thread). */
export const checkForUpdates = (): Promise<UpdateStatus> =>
  ipc.invoke<UpdateStatus>("check_for_updates");

/** Open the releases page (only when the user opts in from the prompt). */
export const openReleasesPage = (): Promise<void> =>
  ipc.invoke("open_releases_page");

/** A newer release was found by the background checker; payload is its tag. */
export const onUpdateAvailable = (handler: (latest: string) => void) =>
  ipc.listen<string>("update:available", handler);

/** Native menu events. */
export const onMenuOpenPack = (handler: () => void) =>
  ipc.listen<null>("menu:open-pack", () => handler());

export const onMenuSave = (handler: () => void) =>
  ipc.listen<null>("menu:save", () => handler());

export const onMenuSaveAs = (handler: () => void) =>
  ipc.listen<null>("menu:save-as", () => handler());

export const onMenuCheckUpdates = (handler: () => void) =>
  ipc.listen<null>("menu:check-updates", () => handler());

export const onMenuClearBoard = (handler: () => void) =>
  ipc.listen<null>("menu:clear-board", () => handler());

export const onMenuPreferences = (handler: () => void) =>
  ipc.listen<null>("menu:preferences", () => handler());

export const onMenuTheme = (handler: (theme: Theme) => void) =>
  ipc.listen<Theme>("menu:theme", handler);

File: src-tauri/src/commands.rs
use std::collections::BTreeMap;

use engine::{PadId, Phase, PlayMode};
use serde::{Deserialize, Serialize};
use tauri::State;

use crate::input::pad_event;
use crate::state::{AppState, AudioCmd, PackEntry, SampleStore, PAD_COUNT};

/// What a `.beat` pack filled: a pad plus the mode the pack assigned it, so the
/// UI can reflect per-pad modes after loading.
#[derive(Serialize)]
pub struct LoadedPad {
    pad: u16,
    mode: String,
}

/// Fixed velocity for keyboard input (the computer keyboard has no velocity
/// sensing). MIDI input will supply real velocities later.
const KEYBOARD_VELOCITY: u8 = 127;

#[tauri::command]
pub fn trigger_pad(pad: u16, state: State<AppState>) {
    if pad >= PAD_COUNT {
        return;
    }
    state.send(AudioCmd::Pad(pad_event(
        pad,
        Phase::Press,
        KEYBOARD_VELOCITY,
    )));
}

#[tauri::command]
pub fn release_pad(pad: u16, state: State<AppState>) {
    if pad >= PAD_COUNT {
        return;
    }
    state.send(AudioCmd::Pad(pad_event(pad, Phase::Release, 0)));
}

#[tauri::command]
pub fn set_pad_mode(pad: u16, mode: String, state: State<AppState>) {
    if pad >= PAD_COUNT {
        return;
    }
    state.send(AudioCmd::SetMode(PadId(pad), parse_mode(&mode)));
}

/// Load (or replace) the sound on a single pad from a file on disk.
#[tauri::command]
pub fn load_pad_sound(
    pad: u16,
    path: String,
    state: State<AppState>,
    samples: State<SampleStore>,
) -> Result<(), String> {
    if pad >= PAD_COUNT {
        return Err("pad out of range".into());
    }
    let bytes = std::fs::read(&path).map_err(|e| e.to_string())?;
    samples.set(pad, file_name(&path), bytes.clone());
    state.send(AudioCmd::LoadSound(PadId(pad), bytes));
    Ok(())
}

/// Load a `.beat` pack, replacing the board. Returns each filled pad + its mode.
#[tauri::command]
pub fn load_beat_pack(
    path: String,
    state: State<AppState>,
    samples: State<SampleStore>,
) -> Result<Vec<LoadedPad>, String> {
    let file = std::fs::File::open(&path).map_err(|e| e.to_string())?;
    let beat = format::read_beat(file).map_err(|e| e.to_string())?;

    let mut entries = Vec::new();
    let mut loaded = Vec::new();
    let mut cached = Vec::new();
    for mapping in &beat.manifest.pads {
        if mapping.pad >= PAD_COUNT {
            log::warn!("pack pad {} is outside the grid; skipped", mapping.pad);
            continue;
        }
        match beat.samples.get(&mapping.sample) {
            Some(bytes) => {
                entries.push(PackEntry {
                    pad: PadId(mapping.pad),
                    bytes: bytes.clone(),
                    mode: from_format_mode(mapping.mode),
                });
                loaded.push(LoadedPad {
                    pad: mapping.pad,
                    mode: format_mode_str(mapping.mode).to_string(),
                });
                cached.push((mapping.pad, mapping.sample.clone(), bytes.clone()));
            }
            None => log::warn!(
                "pack pad {} references missing sample '{}'; skipped",
                mapping.pad,
                mapping.sample
            ),
        }
    }

    // The pack replaces the board, so the save cache is rebuilt to match — only
    // once we know the load is worth applying (a non-empty result).
    if !cached.is_empty() {
        samples.clear();
        for (pad, name, bytes) in cached {
            samples.set(pad, name, bytes);
        }
    }

    state.send(AudioCmd::LoadPack(entries));
    Ok(loaded)
}

/// Stop every running loop at once (panic).
#[tauri::command]
pub fn stop_all(state: State<AppState>) {
    state.send(AudioCmd::StopAll);
}

/// Remove the sound from a single pad.
#[tauri::command]
pub fn clear_pad(pad: u16, state: State<AppState>, samples: State<SampleStore>) {
    if pad >= PAD_COUNT {
        return;
    }
    samples.remove(pad);
    state.send(AudioCmd::Clear(PadId(pad)));
}

/// Manifest fields the UI supplies when saving a board to a `.beat`.
#[derive(Deserialize)]
pub struct SaveMeta {
    name: String,
    author: String,
    bpm: f32,
}

/// Write the current board to a `.beat` archive. `modes` is indexed by pad
/// (the UI owns per-pad modes); sample bytes come from the save cache.
#[tauri::command]
pub fn save_beat(
    path: String,
    meta: SaveMeta,
    modes: Vec<String>,
    samples: State<SampleStore>,
) -> Result<(), String> {
    let store = samples.0.lock().unwrap();
    if store.is_empty() {
        return Err("the board is empty".into());
    }

    let mut sample_bytes = BTreeMap::new();
    let mut pads = Vec::new();
    for (&pad, entry) in store.iter() {
        // Prefix with the pad index so two pads can carry same-named files.
        let sample = format!("{pad}_{}", entry.name);
        sample_bytes.insert(sample.clone(), entry.bytes.clone());
        let mode = modes
            .get(pad as usize)
            .map(|m| to_format_mode(m))
            .unwrap_or(format::PadMode::OneShot);
        pads.push(format::PadMapping { pad, sample, mode });
    }

    let manifest = format::Manifest {
        format_version: format::CURRENT_FORMAT_VERSION,
        name: meta.name,
        author: meta.author,
        bpm: meta.bpm,
        grid: format::GridSize { rows: 8, cols: 8 },
        pads,
        metadata: BTreeMap::new(),
    };
    let beat = format::Beat {
        manifest,
        samples: sample_bytes,
    };
    let file = std::fs::File::create(&path).map_err(|e| e.to_string())?;
    format::write_beat(file, &beat).map_err(|e| e.to_string())
}

/// The trailing file name of a path, for naming samples inside the archive.
fn file_name(path: &str) -> String {
    std::path::Path::new(path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or("sample")
        .to_string()
}

/// Show or hide the native menu bar. The UI owns the visibility state (toggled
/// by Alt) and persists it, so this just applies the requested value.
#[tauri::command]
pub fn set_menu_visible(window: tauri::WebviewWindow, visible: bool) {
    let _ = if visible {
        window.show_menu()
    } else {
        window.hide_menu()
    };
}

fn parse_mode(mode: &str) -> PlayMode {
    match mode {
        "hold_loop" => PlayMode::HoldLoop,
        "toggle_loop" => PlayMode::ToggleLoop,
        _ => PlayMode::OneShot,
    }
}

fn from_format_mode(mode: format::PadMode) -> PlayMode {
    match mode {
        format::PadMode::HoldLoop => PlayMode::HoldLoop,
        format::PadMode::ToggleLoop => PlayMode::ToggleLoop,
        format::PadMode::OneShot => PlayMode::OneShot,
    }
}

fn to_format_mode(mode: &str) -> format::PadMode {
    match mode {
        "hold_loop" => format::PadMode::HoldLoop,
        "toggle_loop" => format::PadMode::ToggleLoop,
        _ => format::PadMode::OneShot,
    }
}

fn format_mode_str(mode: format::PadMode) -> &'static str {
    match mode {
        format::PadMode::OneShot => "one_shot",
        format::PadMode::HoldLoop => "hold_loop",
        format::PadMode::ToggleLoop => "toggle_loop",
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_mode_maps_known_strings_and_defaults_to_one_shot() {
        assert!(matches!(parse_mode("hold_loop"), PlayMode::HoldLoop));
        assert!(matches!(parse_mode("toggle_loop"), PlayMode::ToggleLoop));
        assert!(matches!(parse_mode("one_shot"), PlayMode::OneShot));
        assert!(matches!(parse_mode("nonsense"), PlayMode::OneShot));
    }

    #[test]
    fn to_format_mode_maps_known_strings_and_defaults() {
        assert!(matches!(
            to_format_mode("hold_loop"),
            format::PadMode::HoldLoop
        ));
        assert!(matches!(
            to_format_mode("toggle_loop"),
            format::PadMode::ToggleLoop
        ));
        assert!(matches!(
            to_format_mode("one_shot"),
            format::PadMode::OneShot
        ));
        assert!(matches!(to_format_mode("???"), format::PadMode::OneShot));
    }

    #[test]
    fn format_mode_str_round_trips_through_parse_mode() {
        for mode in [
            format::PadMode::OneShot,
            format::PadMode::HoldLoop,
            format::PadMode::ToggleLoop,
        ] {
            let engine_mode = from_format_mode(mode);
            let reparsed = parse_mode(format_mode_str(mode));
            assert_eq!(
                std::mem::discriminant(&engine_mode),
                std::mem::discriminant(&reparsed)
            );
        }
    }
}
